core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -708,6 +785,15 @@ final class ShuffleBlockFetcherIterator(
   }
 
   private def fetchUpToMaxBytes(): Unit = {
+    if (isNettyOOMOnShuffle.get()) {
+      if (reqsInFlight > 0) {
+        // Return immediately if Netty is still OOMed and there're ongoing fetch requests
+        return
+      } else {
+        ShuffleBlockFetcherIterator.resetNettyOOMFlagIfPossible(0)
+      }
+    }
+
```

Comment:
Was thinking more about this solution and there is a potential problem I see. Once a netty OOM is encountered for some responses, the corresponding requests are deferred and no more remote requests are sent. Now this helps to recover. But we don't change any in-flight remote requests limits. So after the  `isNettyOOMOnShuffle` is reset with a successful remote response, the next burst of remote requests will be sent out at the same rate. This means again there are chances to see netty OOMs and again some of the blocks will be deferred. This introduces more delay and increases the load on shuffle server.

I think solving this maybe more complex and right now this is just a workaround. But maybe we can do something simpler to reduce the number of requests made after a netty OOM is encountered?

Reply:
Yes, that's true. I also considered another way previously, which is to adjust the threshold of in-flight requests dynamically. For example, when OOM throws, the threshold would be reduced to the max number of in-flight requests before OOM. And if OOM happens again, we continue to reduce the threshold. Then, it comes to a question: when do we increase the threshold? When the backlogged requests are too much and OOM has been disappeared for a while? If we go this way, we should be very careful about the adjustment algorithm as it's directly related to the performance.

Let me think more about it. Thanks!

Reply:
So I have limited the reset condition to whether the Netty free memory is larger than `maxReqSizeShuffleToMem` (default 200M), which is more strict than the `averageBlockSize`. I think this would mitigate the issue you mentioned here. WDYT?

Reply:
Should it again check at line 805 that `freeDirectMemory` > maxReqSizeShuffleToMem?
Otherwise it immediately unsets the flag and sends more requests.

Reply:
> Should it again check at line 805 that freeDirectMemory > maxReqSizeShuffleToMem?

No. Otherwise, for the first invocation of `fetchUpToMaxBytes`, the fetching can hang if `isNettyOOMOnShuffle=true.` Because if no requests were sent in the first invocation, there would be no callback on `fetchUpToMaxBytes` later.

Reply:
> Because if no requests were sent in the first invocation, there would be no callback on fetchUpToMaxBytes later.

For the first invocation `fetchUpToMaxBytes` if  `isNettyOOMOnShuffle=true` then that means another iterator saw the NettyOOM as set it to true. This will just unset it.

Also how will it hang? let's just say if they are only remote blocks to be fetched and none of the requests are sent initially, `iterator.next()` will keep calling `fetchUpToMaxBytes`? Eventually when enough freeDirectMemory is available, it will send remote requests.

Reply:
IIUC, it will block at https://github.com/apache/spark/blob/2cb962b132af2683a1a445365636e572c557ac5b/core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala#L594

Reply:
Oh I see. It just waits indefinitely here. Can we not change this to `poll(time)` ? If there is nothing,  result will be null and it will call `fetchUpToMaxBytes` again.
WDYT?

Reply:
`poll(time)` may work but I think it breaks the existing design. After the change, we would introduce the overhead due to call `fetchUpToMaxBytes` many times while it's not necessary.